packages/open-next/src/adapters/server-adapter.ts

```diff
@@ -146,11 +173,23 @@ export async function handler(
       : formatApiv2FailoverResponse();
   }
 
-  // Process Next.js request
+  // Only enable this if you don't need middleware to run on the SSG and ISR request
+  if (process.env.EXPERIMENTAL_CACHE_INTERCEPTION) {
+    //Try intercept the request to see if it is a cached request
+    const cacheResponse = await cacheInterceptor.handler(parser);
+    if (cacheResponse) {
+      // If we have a cache response, return it and don't invoke NextServer
+      return isCloudFrontEvent
+        ? formatCloudFrontResponse(cacheResponse)
+        : formatApiv2Response(cacheResponse);
+    }
+  }
+
   const reqProps = {
     method: parser.method,
     url: parser.url,
-    headers: parser.headers,
+    //WORKAROUND: We pass this header to the serverless function to mimic a prefetch request which will not trigger revalidation since we handle revalidation differently
+    headers: { ...parser.headers, purpose: "prefetch" },
```

Comment:
We don't override `ResponseCache` but `IncrementalCache` : https://github.com/vercel/next.js/blob/3003bff6fa921c1df851d3e442beba9f1e5c65ef/packages/next/src/server/lib/incremental-cache/file-system-cache.ts
So i don't see how we could use the settings that you've pointed

Reply:
> So i don't see how we could use the settings that you've pointed

Clearly I got lost looking for the right use case, very helpful, thank you.

Reply:
> I believe if we did so, it would somehow involve the params handed into `get` in `cache.ts`. Currently there is a boolean for second parameter, but I see a function in the normal `ResponseCache` in next.js: https://github.com/vercel/next.js/blob/3003bff6fa921c1df851d3e442beba9f1e5c65ef/packages/next/src/server/response-cache/index.ts#L28

This got me thinking of another way to skip revalidation. We could calculate if data is stale directly into `get` of our `cache.ts`, and if that's the case set the `lastModified` date of our cache entry to `Date.now()` and trigger a background revalidation. Next would think that our data is fresh and so return directly without trying to revalidate.

One of the downsides of this solution is that next will set cache-control headers as `s-maxage=X stale-while-revalidate` and the cdn will cache the stale data for as long as revalidate.

Not sure if that's the way to go

Reply:
> One of the downsides of this solution is that next will set cache-control headers as s-maxage=X stale-while-revalidate and the cdn will cache the stale data for as long as revalidate.

This is true, and no amount of logic will stop that from occurring if we trick Next, but your idea is a much more accurate implementation of what I had in my head to be sure.

Feels like it would be so beautiful to somehow inject our own `revalidate` similar to how we have taken over `cache`, and have that pop out an SQS message instead of a blocking build.

Reply:
Does `prefetch` not suffer from the same cache-control header problem?

Looking at this line, which appears to treat a non-stale page identically to a prefetch = true: https://github.com/vercel/next.js/blob/3003bff6fa921c1df851d3e442beba9f1e5c65ef/packages/next/src/server/response-cache/index.ts#L117

Reply:
You're right, but i added this line which solved the issue : https://github.com/conico974/open-next/blob/b8be6f93ca64b6ae928c34d2e716f717cd1d1a39/packages/open-next/src/adapters/server-adapter.ts#L222

We won't be able to do that with the other solution since `x-nextjs-cache` headers would be `HIT`

Reply:
Am I understanding correctly:
Without any of your changes, the NextServer would:
- Try to respond with the stale page.
- Background a revalidation request, blocking lambda from responding.
- Finish the revalidation request, save to disk.
- Respond with the stale page.
?

OR

Is it the same, but it responds with the NEW page?

Reply:
It's even worse than that it : 
- Try to respond with the stale page
- Launch revalidation in the background but don't block anything
- Return the stale page
- Here depending on the speed at which the revalidation happens it could either : 
  - If revalidation is fast enough try to save to disk, fails because this part of the disk is read-only, and resort to using memory cache
  - Otherwise it will potentially finish the revalidation once the lambda receive a new request, but there is absolutely no guarantee since we have already returned here. Sometimes it might work if the subsequent request to the same lambda are close enough

Reply:
Thinking this through:
If you were to only replace the cache, and leave everything else as is, it should in theory:

- Launch revalidation in background
- Return stale page
- No longer fail, write revalidation to S3
- Next time respond from updated S3?

Reply:
That's what i tested first, but it leads to some inconsistent result.
- Sometimes the revalidation would just not finish at all, for example when hitting a lambda that is never touched again
- I got some write error to S3 since the lambda could be stopped in the middle of a write
- I also got the case where only the html has been saved to S3 but not the json. This one could be a pretty bad experience for the user since they could see fresh data first then on navigation get stale data for the same page

We could potentially remove the prefetch headers and let the lambda trying to revalidate in the background. But i think to be certain that it works correctly, we shoud at least try to revalidate ourselves.
If we do that, i can see 2 things happening:
- We would potentially revalidate data twice or more
- We could get into a very rare edge case where : 
  - User access page A on lambda 1
  - Page A is starting to revalidate on lambda 1, but stopped since lambda 1 already returned
  - We revalidate ourselves 
  - A new request hit lambda 1 and it resumes execution of the background revalidation
  - The request returned on lambda 1 so it stopped the background revalidation again and wrote only the html data
  - Here we get an inconsistent state

Honestly i think these tradeoff are perfectly acceptable

Reply:
That sounds like the expected race condition from lambda, where the call may or may not complete before the response.  Was worth asking, but I figured you had encountered this.

This part of the implementation is the only part that feels like it would be different from the way Vercel handles it.  That being said, after being bathed in Vercel code yesterday, I am actually fully capable of believing that they would use this type of workaround permanently.

Thanks for answering all of my questions.

Reply:
No problem, thank you for taking the time to review

I just updated the PR but I didn't change this part.
I added some comment in the code to explain why we chose prefetch and suggested some alternative solutions.

Please let me know if you have any further questions.